### ISSUE TITLE
Process exits with SIGTRAP after doing vfork + exec under kcov

### ISSUE THREAD

**Issue by VincentGibert** [SimonKagstrom/kcov#110]
Hi,

Thanks to your fix for the issue #104, I'm now able to cover properly the net-snmp daemon. However, I'm facing some issues when doing code coverage of a custom snmp module (dynamically loaded) using kcov.

Here is an overview of the module internals:

 - at initialization, the module spawns a  bunch of threads (one for each subsystem to monitor),
 - each thread spawns a few subthreads (one for each subsystem component),
 - each sub-thread gets the state of the component by executing a system command with a custom implementation of popen (popen provided by the glibc is not threadsafe). The custom popen function spawns a child using vfork and exec,
 - the value is stored in a memory area and the main snmpd thread just has to read the memory when the value is requested by a snmp get request for example.

When running the snmpd daemon with kcov, I've noticed the following behaviour:

    $ kcov /tmp/kcov_results /usr/sbin/snmpd -f

--> The report generated by kcov includes coverage of the net-snmp code but the custom module code is not covered. There are no visible tasks in /proc/\<snmpd_pid\>/task and I can't see any trace from the module in syslog. It seems that the code of the module is not executed.

    $ kcov --include-path=/path/to/custom/module/sourcedir /tmp/kcov_results /usr/sbin/snmpd -f
    kcov: Process exited with signal 5 (unknown signal 5) at 0x0

--> The report generated by kcov includes coverage of the custom module and doesn't includes coverage of net-snmp code anymore. The creation of the threads is correctly logged by the module. However, the program crashes quickly by receiving the signal 5 (SIGTRAP). According to the coredump, it seems the signal occurs just after a sub-thread executes vfork.

    $ kcov --include-path=/path/to/custom/module/sourcedir /tmp/kcov_results /usr/sbin/snmpd

--> Without the `-f` flag, the main thread of snmpd fork. Hence the threads of the custom module die when the parent exits. However, the child continues to execute a part of the module code and it is correctly covered by kcov. Of course, the part of the code belonging to the threads has a coverage of 0% while the part executed by the main thread has a better score.

    $ kcov /tmp/kcov_results /usr/sbin/snmpd

--> In this case, the report generated by kcov includes only coverage from the net-snmp source code and the custom module is not covered at all.

Is it possible to cover such a corner case (thread + thread + fork in a dlopened library) with kcov ?

**Comment by SimonKagstrom** [comment]
Well, dlopens should be handled. Kcov uses a trick with LD_PRELOAD to handle these (and other shared libraries) - providing it's own dlopen() call to know when a shared library is added, but that should work fine.

Apparently there is some problem in this case though. Could you try the dlopen example in the kcov distribution on your system? It's placed in tests/ and you can build that and all other tests by running cmake/make in the tests subdir.

**Comment by SimonKagstrom** [comment]
Another random thought: Could you try replacing the vfork with a fork?

**Comment by VincentGibert** [comment]
I ran your dlopen test and it seems ok:

    $ LD_LIBRARY_PATH=. kcov /tmp/kcov_results ./dlopen
    In shared library
    from shared lib: 1

Coverage results from this test looks fine to me.

**Comment by VincentGibert** [comment]
When replacing vfork by fork in the custom popen function I don't get SIGTRAP anymore, however I get a SIGSEGV signal. Without kcov, the SIGSEGV isn't raised. I'm wondering if a race condition occurs in the module when it is traced by kcov. I will try to investigate further.

**Comment by SimonKagstrom** [comment]
OK, thanks for testing.

So basically your program does something like

```
my_popen(...)
{
   pid = vfork();
   if (pid == 0)
   {
       exec(...);
   }
   ...
}
```

I guess I should do some testing with vfork/exec as well (I don't see why there should be a difference though). If you run kcov with --debug=15, does it appear to SIGSEGV/SIGTRAP in the same place with fork/vfork?

kcov affects performance a lot (as you've probably seen), so race conditions can certainly be triggered in some cases. I've been hit myself with that kind of problems. However, I don't think it's the most likely explanation, but perhaps you can get more hints by seeing where it segfaults.

**Comment by VincentGibert** [comment]
I have done some tests using `--debug=15`

- **with fork:** debug traces show that kcov is trying to follow the exec'ed process. For instance, kcov searches debug symbols in a library which is loaded by the new process (the new process is nvidia-settings and kcov searches symbols in libnvidia-glcore.so.346.47). The SIGSEGV signal doesn't occur at the same point for each run, so it raises my suspicion about a race condition in the custom module.

- **with vfork:** debug traces doesn't show the search of the symbols in nvidia's shared library. It seems that the exec() syscall is not reached. It looks like the SIGTRAP signal occurs always at the same time, just after the vfork call.

So it seems fork and vfork are handled differently by kcov.

**Comment by VincentGibert** [comment]
After further investigation, I managed to find the race condition in the module. With the use of fork instead of vfork and the fix of the race condition kcov seems to work nicely.

**Comment by SimonKagstrom** [comment]
Thanks for the investigation and great work to find the race condition!

I'm changing the title to reflect the vfork issue then.

**Comment by SimonKagstrom** [comment]
I've pushed a potential fix for the vfork issue (blindly), care to test?

**Comment by VincentGibert** [comment]
Unfortunately I don't see any change with this fix.

**Comment by VincentGibert** [comment]
I'm not sure if it is relevant: using your test case for fork (`tests/fork/fork.c`), I replaced all `fork()` calls by `vfork()`. The resulting program also exits with SIGTRAP.

    $ kcov --debug=15 /tmp/kcov_results vfork
    PT forked 26515
    BP registered at 0x400864
    REPORT /usr/src/kcov-28.20150730.gitdf8ea16/tests/fork/vfork.c:10 at 0x400864
    REPORT /usr/src/kcov-28.20150730.gitdf8ea16/tests/fork/vfork.c:10 at 0xa90a56384
    REPORT /usr/src/kcov-28.20150730.gitdf8ea16/tests/fork/vfork.c:10 at 0xa90a56384
    ...
    BP registered at 0x4009a5
    REPORT /usr/src/kcov-28.20150730.gitdf8ea16/tests/fork/vfork.c:47 at 0x4009a5
    REPORT /usr/src/kcov-28.20150730.gitdf8ea16/tests/fork/vfork.c:47 at 0x2f90a56384
    REPORT /usr/src/kcov-28.20150730.gitdf8ea16/tests/fork/vfork.c:47 at 0x2f90a56384
    PT continuing 26515 with signal 0
    PT stopped PID 26515 0x0000057f
    PT BP at 0x7ffff77dbd19:-1 for 26515
    Can't find breakpoint at 0x7ffff77dbd19
    PT continuing 26515 with signal 0
    PT stopped PID 26515 0x0000057f
    PT BP at 0x400874:-1 for 26515
    REPORT hit at 0x400874
    REPORT hit at 0xf90a56384
    REPORT hit at 0xf90a56384
    PT continuing 26515 with signal 0
    PT stopped PID 26515 0x0000057f
    PT BP at 0x400883:-1 for 26515
    REPORT hit at 0x400883
    REPORT hit at 0x1390a56384
    REPORT hit at 0x1390a56384
    PT continuing 26515 with signal 0
    PT stopped PID 26515 0x00000005
    PT terminating signal 5 at 0x0 for 26515
    PT continuing 26515 with signal 0
    PT error for 26515: -1
    Returning error
    REPORT /usr/src/kcov-28.20150730.gitdf8ea16/tests/fork/vfork.c:46 at 0x2e90a56384
    REPORT /usr/src/kcov-28.20150730.gitdf8ea16/tests/fork/vfork.c:47 at 0x2f90a56384
    REPORT /usr/src/kcov-28.20150730.gitdf8ea16/tests/fork/vfork.c:25 at 0x1990a56384
    ...
    kcov: Process exited with signal 5 (unknown signal 5) at 0x0

**Comment by SimonKagstrom** [comment]
Thanks for the test! With a simple case like that it should be easier to find some clues. I'll be offline for a couple of days, but will take a look at this again thereafter.

**Comment by SimonKagstrom** [comment]
Fixed with e387cee. Sorry for taking so long for the fix!

**Comment by VincentGibert** [comment]
I just tried with this fix and the issue is fixed now.

Thanks a lot !
